### PR TITLE
Add support for STM32duino DFU bootloader.

### PR DESCRIPTION
**Pull request by tzarc**
## Description

Adds support for the STM32duino bootloader, VID=0x1eaf, PID=0x0003.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Documentation

### CODE REVIEW

**Comment by tzarc** [comment]
I'm not... but I can probably have a stab.

**Comment by nooges** [comment]
@tzarc ping/tag me if needed regarding obj-c

File: windows/QMK Toolbox/MainWindow.cs [hunk 1]
```diff
@@ -201,7 +201,7 @@ private void MainWindow_Load(object sender, EventArgs e)
             _printer.PrintResponse(" - Atmel/LUFA/QMK DFU via dfu-programmer (http://dfu-programmer.github.io/)\n", MessageType.Info);
             _printer.PrintResponse(" - Caterina (Arduino, Pro Micro) via avrdude (http://nongnu.org/avrdude/)\n", MessageType.Info);
             _printer.PrintResponse(" - Halfkay (Teensy, Ergodox EZ) via Teensy Loader (https://pjrc.com/teensy/loader_cli.html)\n", MessageType.Info);
-            _printer.PrintResponse(" - ARM DFU (STM32, Kiibohd) via dfu-util (http://dfu-util.sourceforge.net/)\n", MessageType.Info);
+            _printer.PrintResponse(" - ARM DFU (STM32, Kiibohd, STM32duino) via dfu-util (http://dfu-util.sourceforge.net/)\n", MessageType.Info);
```
Comment (fauxpark) [inline windows/QMK Toolbox/MainWindow.cs]: Should also be added to Program.cs, and the readme. For macOS, main.m and AppDelegate.m.

File: windows/QMK Toolbox/USB.cs [hunk 2]
```diff
@@ -104,6 +104,11 @@ public bool DetectBootloader(ManagementBaseObject instance, bool connected = tru
                 deviceName = "BootloadHID";
                 _devicesAvailable[(int)Chipset.BootloadHid] += connected ? 1 : -1;
             }
+            else if (MatchVid(deviceId, 0x1eaf) && MatchPid(deviceId, 0x0003)) // Detects STM32duino Bootloader
```
Comment (fauxpark) [inline windows/QMK Toolbox/USB.cs]:
```suggestion
            else if (MatchVid(deviceId, 0x1EAF) && MatchPid(deviceId, 0x0003)) // Detects STM32duino Bootloader
```

**Comment by fauxpark** [comment]
Do we also need to add this bootloader's VID/PID to drivers.txt?

**Comment by tzarc** [comment]
> 
> 
> Do we also need to add this bootloader's VID/PID to drivers.txt?

If that's to support Zadig, I don't believe it's needed..., I certainly didn't need to install anything other than the default that came up.

I'm actually thinking that perhaps this isn't a PR that should be merged -- in order to get uploads to work with the STM32duino bootloader I've been finding that "auto flash" needs to be turned on in toolbox. By the time it shows up as a new DFU bootloader device in toolbox it's already too late, as it's already started executing any user code that was already on the board.

Given the small number of people targeting this bootloader (are there any?) I'm inclined to just close without merge. Thoughts?

**Comment by fauxpark** [comment]
Looks like libusb-win32:
https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/drivers/win/src/libwdi-1.2.4-with-usbser-driver-v3.diff#L9
https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/master/drivers/win/install_drivers.bat#L4

As for the auto flash being necessary, it doesn't sound any worse than Caterina. I'm not averse to merging if that's the case. We can probably also optimise the whole bootloader detection process, to speed it up, though that's easier said than done 😄

**Comment by awkannan** [comment]
> > Do we also need to add this bootloader's VID/PID to drivers.txt?
> 
> If that's to support Zadig, I don't believe it's needed..., I certainly didn't need to install anything other than the default that came up.
> 
> I'm actually thinking that perhaps this isn't a PR that should be merged -- in order to get uploads to work with the STM32duino bootloader I've been finding that "auto flash" needs to be turned on in toolbox. By the time it shows up as a new DFU bootloader device in toolbox it's already too late, as it's already started executing any user code that was already on the board.
> 
> Given the small number of people targeting this bootloader (are there any?) I'm inclined to just close without merge. Thoughts?

Just gonna add my 2c - the stm32duino bootloader does support a boot pin that can be held high to put it into bootloader mode. For this reason, I think it's not a bad idea to merge this.

Also, autoflash is a viable option!

**Comment by fauxpark** [comment]
> Just gonna add my 2c - the stm32duino bootloader does support a boot pin that can be held high to put it into bootloader mode. For this reason, I think it's not a bad idea to merge this.

Aha! I was trying to figure out why I wasn't experiencing the immediate jump to user code tzarc mentioned above. Turns out I completely missed the `binaries` directory in the repo and flashed the one in `bootloader_only_binaries`. Sure enough it's now a serial device and won't stay in bootloader when I hit the reset button, but if I flip the BOOT1 jumper, it does.

I'm just waiting on a rebase of this PR and for the reviews to be applied, as well as adding the device to drivers.txt, then I think it can be merged.

File: windows/QMK Toolbox/Flashing.cs [hunk 3]
```diff
@@ -230,6 +233,15 @@ private void FlashStm32Dfu(string mcu, string file)
             }
         }
 
+        private void FlashStm32Duino(string file)
+        {
+            if (Path.GetExtension(file)?.ToLower() == ".bin") {
+                RunProcess("dfu-util.exe", $"-a 2 -d 1eaf:0003 -R -D \"{file}\"");
```
Comment (fauxpark) [inline windows/QMK Toolbox/Flashing.cs:239]:
```suggestion
                RunProcess("dfu-util.exe", $"-a 2 -d 1EAF:0003 -R -D \"{file}\"");

```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**